File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool ValidBST(TreeNode* root, long left, long right){
        if(!root){
            return true;
        }
        if(!(left < root->val && right > root->val)){
           return false; 
        }
        return ValidBST(root->left, left, root->val) && ValidBST( root->right, root->val, right);
    }
    bool isValidBST(TreeNode* root) {
        return ValidBST( root, LONG_MIN, LONG_MAX);
    }
};

File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        res = 0

        left = 0
        right = len(heights) - 1

        while left < right:
            if heights[right] > heights[left]:
                area =  heights[left] * (right - left)
                if area > res:
                    res = area 
                left += 1 
            elif heights[right] < heights[left]:
                area =  heights[right] * (right - left)
                if area > res:
                    res = area 
                right -= 1 
            else:
                area =  heights[right] * (right - left)
                if area > res:
                    res = area 
                if heights[right - 1] > heights[left + 1]:
                    right -= 1 
                else:
                    left += 1

        return res